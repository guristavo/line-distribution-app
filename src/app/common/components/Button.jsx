import React, { Fragment } from 'react';
import PropTypes from 'prop-types';

// Import shared components
import { Icon } from '../../common';
// Import utility functions
import { bem } from '../../../utils';

const Button = ({
  label,
  onClick,
  className,
  icon,
  isBlock,
  isDisabled,
  isVisible,
  size,
  type,
  children,
}) => {
  const blockClass = isBlock ? 'block' : '';
  const visibilityClass = isVisible ? '' : ' hidden';
  const bemClasses = bem('btn', [blockClass, size, type]);

  return (
    <button
      className={`${bemClasses} ${className} ${visibilityClass}`}
      onClick={() => onClick()}
      disabled={isDisabled}
    >
      {children ? (
        <Fragment>{children}</Fragment>
      ) : (
        <Fragment>
          {Boolean(icon) && <Icon type={icon} size="12" inline />}
          {label}
        </Fragment>
      )}
    </button>
  );
};

Button.defaultProps = {
  className: '',
  icon: '',
  isBlock: false,
  isDisabled: false,
  isVisible: true,
  size: 'default',
  type: 'primary',
  children: null,
};

Button.propTypes = {
  label: PropTypes.string.isRequired,
  onClick: PropTypes.func.isRequired,
  className: PropTypes.string,
  icon: PropTypes.string,
  isBlock: PropTypes.bool,
  isDisabled: PropTypes.bool,
  isVisible: PropTypes.bool,
  size: PropTypes.oneOf(['small', 'default', 'large']),
  type: PropTypes.oneOf(['primary', 'secondary', 'terciary', 'danger']),
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node,
  ]),
};

export default Button;
