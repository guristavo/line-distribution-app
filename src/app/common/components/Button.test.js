import React from 'react';
import { render, fireEvent, cleanup } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import Button from './Button';

describe('Common/Button', () => {
  let NOOP;
  const FALSE = false;
  const TRUE = true;

  beforeAll(() => {});

  beforeEach(() => {
    NOOP = jest.fn();
  });

  afterEach(cleanup);

  it('renders', () => {
    const { asFragment } = render(<Button label="Label" onClick={NOOP} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders children as a block', () => {
    const { getByText, asFragment } = render(
      <Button label="Label" onClick={NOOP}>
        <p>Button Name Here</p>
      </Button>
    );
    expect(asFragment()).toMatchSnapshot();
    expect(getByText('Button Name Here')).toBeInTheDocument();
  });

  it('requires label prop', () => {
    const { getByText } = render(<Button label="hello" onClick={NOOP} />);

    expect(getByText('hello')).toBeInTheDocument();

    const { getByText: getByText2 } = render(
      <Button label="world" onClick={NOOP} />
    );

    expect(getByText2('world')).toBeInTheDocument();
  });

  it('fires onClick prop function when clicked', () => {
    const { getByText } = render(<Button label="Label" onClick={NOOP} />);

    expect(NOOP).not.toHaveBeenCalled();

    fireEvent.click(getByText('Label'));

    expect(NOOP).toHaveBeenCalled();
  });

  it('accepts a icon name as prop', () => {
    const { getByTitle } = render(
      <Button label="Label" onClick={NOOP} icon="lock" />
    );

    expect(getByTitle('lock')).toBeInTheDocument();
  });

  it('has default classes', () => {
    const { container } = render(<Button label="Label" onClick={NOOP} />);

    expect(container.firstChild.classList.value.trim()).toBe(
      'btn btn--default btn--primary'
    );
  });

  it('adds an block class when isBlock flag is true', () => {
    const { container } = render(
      <Button label="Label" onClick={NOOP} isBlock={TRUE} />
    );

    expect(container.firstChild.classList.contains('btn--block')).toBe(true);
  });

  it('adds an hidden class when isVisible flag is false', () => {
    const { container } = render(
      <Button label="Label" onClick={NOOP} isVisible={FALSE} />
    );

    expect(container.firstChild.classList.contains('hidden')).toBe(true);
  });

  it('adds btn--default when size argument is "default" or omitted', () => {
    const { container } = render(
      <Button label="Label" onClick={NOOP} size="default" />
    );

    expect(container.firstChild.classList.contains('btn--default')).toBe(true);
  });

  it('adds btn--small class when size argument is "small"', () => {
    const { container } = render(
      <Button label="Label" onClick={NOOP} size="small" />
    );

    expect(container.firstChild.classList.contains('btn--small')).toBe(true);
  });

  it('adds btn--large class when size argument is "large"', () => {
    const { container } = render(
      <Button label="Label" onClick={NOOP} size="large" />
    );

    expect(container.firstChild.classList.contains('btn--large')).toBe(true);
  });

  it('adds btn--primary class when type argument is "primary" or omitted', () => {
    const { container } = render(
      <Button label="Label" onClick={NOOP} type="primary" />
    );

    expect(container.firstChild.classList.contains('btn--primary')).toBe(true);
  });

  it('adds btn--secondary class when type argument is "secondary"', () => {
    const { container } = render(
      <Button label="Label" onClick={NOOP} type="secondary" />
    );

    expect(container.firstChild.classList.contains('btn--secondary')).toBe(
      true
    );
  });

  it('adds btn--terciary class when type argument is "terciary"', () => {
    const { container } = render(
      <Button label="Label" onClick={NOOP} type="terciary" />
    );

    expect(container.firstChild.classList.contains('btn--terciary')).toBe(true);
  });

  it('adds btn--danger class when type argument is "danger"', () => {
    const { container } = render(
      <Button label="Label" onClick={NOOP} type="danger" />
    );

    expect(container.firstChild.classList.contains('btn--danger')).toBe(true);
  });

  it('accepts custom classes with the className argument"', () => {
    const { container } = render(
      <Button
        label="Label"
        onClick={NOOP}
        className="custom-class extra-class"
      />
    );

    expect(container.firstChild.classList.contains('btn')).toBe(true);
    expect(container.firstChild.classList.contains('custom-class')).toBe(true);
    expect(container.firstChild.classList.contains('extra-class')).toBe(true);
  });
});
