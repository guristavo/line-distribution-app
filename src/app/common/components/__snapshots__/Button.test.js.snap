// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Common/Button renders 1`] = `
<DocumentFragment>
  <button
    class="btn btn--default btn--primary  "
  >
    Label
  </button>
</DocumentFragment>
`;

exports[`Common/Button renders children as a block 1`] = `
<DocumentFragment>
  <button
    class="btn btn--default btn--primary  "
  >
    <p>
      Button Name Here
    </p>
  </button>
</DocumentFragment>
`;
